include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h

```diff
@@ -3461,6 +3514,373 @@ MGTransferBlockGlobalCoarsening<dim, VectorType>::get_matrix_free_transfer(
+            [](const auto &a, const auto &b) { return a.first < b.first; });
+  points_all.erase(std::unique(points_all.begin(),
+                               points_all.end(),
+                               [](const auto &a, const auto &b) {
+                                 return a.first == b.first;
+                               }),
+                   points_all.end());
+
+  for (const auto &i : points_all)
+    {
+      this->level_dof_indices_fine.push_back(i.first);
+      points.push_back(i.second);
+    }
+
+  // Duplicates support points have been removed, hand them over to rpe.
+  rpe.reinit(points, dof_handler_coarse.get_triangulation(), mapping_coarse);
+
+  // set up MappingInfo for easier data access
+  mapping_info = internal::fill_mapping_info(rpe);
+
+  // set up constraints
+  const auto &cell_data = rpe.get_cell_data();
+
+  constraint_info.reinit(dof_handler_coarse,
+                         cell_data.cells.size(),
+                         false /*TODO*/);
+
+  for (unsigned int i = 0; i < cell_data.cells.size(); ++i)
+    {
+      typename DoFHandler<dim>::active_cell_iterator cell = {
+        &rpe.get_triangulation(),
+        cell_data.cells[i].first,
+        cell_data.cells[i].second,
+        &dof_handler_coarse};
+
+      constraint_info.read_dof_indices(i,
+                                       numbers::invalid_unsigned_int,
+                                       cell,
+                                       constraint_coarse,
+                                       this->partitioner_coarse);
+    }
+
+  constraint_info.finalize();
+
+  const auto &fe_base = dof_handler_coarse.get_fe().base_element(0);
+
+  if (const auto fe = dynamic_cast<const FE_Q<dim> *>(&fe_base))
+    fe_coarse = std::make_unique<FE_DGQ<dim>>(fe->get_degree());
+  else if (const auto fe = dynamic_cast<const FE_DGQ<dim> *>(&fe_base))
+    fe_coarse = fe->clone();
+  else
+    AssertThrow(false, ExcMessage(dof_handler_coarse.get_fe().get_name()));
+}
+
+
+
+template <int dim, typename Number>
+void
+MGTwoLevelTransferNonNested<dim, LinearAlgebra::distributed::Vector<Number>>::
+  prolongate_and_add_internal(
+    LinearAlgebra::distributed::Vector<Number> &      dst,
+    const LinearAlgebra::distributed::Vector<Number> &src) const
+{
+  std::vector<Number> evaluation_point_results;
+  std::vector<Number> buffer;
+
+  const auto evaluation_function = [&](auto &values, const auto &cell_data) {
+    std::vector<Number> solution_values;
+
+    FEPointEvaluation<1, dim, dim, Number> evaluator(*mapping_info, *fe_coarse);
+
+    for (unsigned int cell = 0; cell < cell_data.cells.size(); ++cell)
+      {
+        solution_values.resize(fe_coarse->n_dofs_per_cell());
+
+        // gather and resolve constraints
+        internal::VectorReader<Number, VectorizedArrayType> reader;
+        constraint_info.read_write_operation(
+          reader,
+          src,
+          reinterpret_cast<VectorizedArrayType *>(solution_values.data()),
```

Comment:
Don't we support plain (non-vectorized) read access in `ConstraintInfo`? I guess this relates to the newly opened issue #15162.